src/crimson/common/errorator.h

```diff
@@ -604,6 +604,10 @@ struct errorator {
 
     template <class...>
     friend class ::seastar::future;
+
+    // let seastar::do_with to up-cast us to seastar::future.
+    template<typename T, typename F>
+    friend inline auto ::seastar::do_with(T&&, F&&);
```

Comment:
i don't really understand this. by making `::seastar::do_with(T&&, F&&)`, this function still cannot cast `_future` to its private base class.

Reply:
`seastar::do_with()` internally calls `seastar::internal::set_callback()` which takes `seastar::future` as it first parameter. Making `do_with` our friend actually allows for the implicit up-cast conversion to take place there.

Reply:
```c++
struct foo_t {
  int i;
};

struct bar_t : private foo_t {
  bar_t() = default;
private:
  friend void func(foo_t&);
};

void func(foo_t& f) {
  
}

int main()
{
  bar_t b;
  func(b);
}
```

does not compile. see https://godbolt.org/z/QNStKP

Reply:
That's not the case. Let me mimic `do_with` with the following snippet:

```cpp
struct foo_t {
  int i;
};

struct bar_t : private foo_t {
  bar_t() = default;
private:
  friend void barfunc(bar_t&);
};

void foofunc(foo_t& f) {

}
void barfunc(bar_t& f) {
  foofunc(f);
}

int main()
{
  bar_t b;
  barfunc(b);
}
```

https://godbolt.org/z/xbgLas